src/parser.rs

```diff
@@ -0,0 +1,200 @@
+use crate::expr::{Expr, Unary};
+use crate::parse_error::ParseError;
+use crate::token::token::Token;
+use crate::token::token_type::TokenType;
+
+#[derive(Default)]
+pub struct Parser {
+    pub tokens: Vec<Token>,
+    pub current: usize,
+}
+
+impl Parser {
+    fn expression(self) -> Expr {
+        return self.equality();
+    }
+
+    pub fn parse(self) -> u8 {
+        todo!();
+    }
+
+    fn synchronize(mut self) {
+        self.advance();
+
+        while !self.is_at_end() {
+            if self.previous().unwrap().token_type == TokenType::SEMICOLON {
+                return;
+            }
+
+            if matches!(
+                &self.peek().unwrap().token_type,
+                TokenType::CLASS
+                    | TokenType::FUN
+                    | TokenType::VAR
+                    | TokenType::FOR
+                    | TokenType::IF
+                    | TokenType::WHILE
+                    | TokenType::PRINT
+                    | TokenType::RETURN
+            ) {
+                return;
+            }
+
+            self.advance();
+        }
+    }
+
+    fn consume(&mut self, token_type: TokenType, message: String) -> Result<&Token, ParseError> {
+        if self.check(token_type) {
+            return Ok(self.advance().unwrap());
+        }
+
+        let peek = self.peek().unwrap();
+
+        return Err(ParseError { peek, message });
+    }
+
+    fn primary(&self) -> Expr {
+        if (self.match_types(vec![TokenType::FALSE])) {
+            return Expr::Literal(Some(false));
+        }
+        if (self.match_types(vec![TokenType::TRUE])) {
+            return Expr::Literal(Some(true));
+        }
+        if (self.match_types(vec![TokenType::NIL])) {
+            return Expr::Literal(None);
+        }
+
+        if (self.match_types(vec![TokenType::NUMBER, TokenType::STRING])) {
+            return Expr::Literal(self.previous().unwrap().literal);
+        }
+
+        if (self.match_types(vec![TokenType::LEFT_PAREN])) {
+            let expr = self.expression();
+            self.consume(
+                TokenType::RIGHT_PAREN,
+                String::from("Expect ')' after expression."),
+            );
+            return Expr::Grouping(expr);
+        }
+    }
+
+    fn unary(self) -> Unary {
+        if self.match_types(vec![TokenType::BANG, TokenType::MINUS]) {
+            let operator = self.previous();
+            let right = self.unary();
+            return Unary {
+                operator,
+                right: Some(Box::new(right)),
+            };
+        }
```

Comment:
@keithamus; Any idea on how I should represent the `Expr` -> `Unary` super class -> sub class relationship? The book just subclasses because it's in Java but I'm not sure what the equivalent would be in Rust.

Reply:
The easiest way, if you have a known set of "subclasses" is to group them all into an enum:

```rust
enum Expr {
  Unary { operand: Box<Expr>, operator: Token },
  Binary { left_operand: Box<Expr>, operator: Token, right_operand: Box<Expr> },
}
```

Of course this will require you to match against each type for any method where you might return data, but that's likely okay.